src/api.h

```diff
@@ -39,25 +34,325 @@ AWS_CRT_API const char *aws_crt_error_debug_str(int);
 
 /* Auth */
-AWS_CRT_API struct aws_credentials *aws_crt_credentials_new(
-    const char *access_key_id, const char *secret_access_key,
-    const char *session_token, uint64_t expiration_timepoint_seconds);
-AWS_CRT_API void
-aws_crt_credentials_release(struct aws_credentials *credentials);
-AWS_CRT_API struct aws_byte_cursor aws_crt_credentials_get_access_key_id(
-    const struct aws_credentials *credentials);
-AWS_CRT_API struct aws_byte_cursor aws_crt_credentials_get_secret_access_key(
-    const struct aws_credentials *credentials);
-AWS_CRT_API struct aws_byte_cursor aws_crt_credentials_get_session_token(
-    const struct aws_credentials *credentials);
-AWS_CRT_API uint64_t aws_crt_credentials_get_expiration_timepoint_seconds(
-    const struct aws_credentials *credentials);
+typedef struct aws_credentials aws_crt_credentials;
+typedef struct _aws_crt_credentials_options aws_crt_credentials_options;
+AWS_CRT_API aws_crt_credentials_options *aws_crt_credentials_options_new(void);
+AWS_CRT_API void aws_crt_credentials_options_release(aws_crt_credentials_options *options);
+AWS_CRT_API void aws_crt_credentials_options_set_access_key_id(
+    aws_crt_credentials_options *options,
+    const char *access_key_id);
+AWS_CRT_API void aws_crt_credentials_options_set_secret_access_key(
+    aws_crt_credentials_options *options,
+    const char *secret_access_key);
+AWS_CRT_API void aws_crt_credentials_options_set_session_token(
+    aws_crt_credentials_options *options,
+    const char *session_token);
+AWS_CRT_API void aws_crt_credentials_options_set_expiration_timepoint_seconds(
+    aws_crt_credentials_options *options,
+    uint64_t expiration_timepoint_seconds);
+AWS_CRT_API aws_crt_credentials *aws_crt_credentials_new(aws_crt_credentials_options *options);
+AWS_CRT_API void aws_crt_credentials_release(aws_crt_credentials *credentials);
+AWS_CRT_API struct aws_byte_cursor aws_crt_credentials_get_access_key_id(const aws_crt_credentials *credentials);
+AWS_CRT_API struct aws_byte_cursor aws_crt_credentials_get_secret_access_key(const aws_crt_credentials *credentials);
+AWS_CRT_API struct aws_byte_cursor aws_crt_credentials_get_session_token(const aws_crt_credentials *credentials);
+AWS_CRT_API uint64_t aws_crt_credentials_get_expiration_timepoint_seconds(const struct aws_credentials *credentials);
+
+/* Credentials providers */
+typedef struct aws_credentials_provider aws_crt_credentials_provider;
+
+/* static credentials provider */
+typedef struct _aws_crt_credentials_provider_static_options aws_crt_credentials_provider_static_options;
+AWS_CRT_API aws_crt_credentials_provider_static_options *aws_crt_credentials_provider_static_options_new(void);
+AWS_CRT_API void aws_crt_credentials_provider_options_static_release(
+    aws_crt_credentials_provider_static_options *options);
+AWS_CRT_API const char *aws_crt_credentials_provider_static_options_get_access_key_id(
+    aws_crt_credentials_provider_static_options *options);
```

Comment:
we probably don't need getters for options-structs

Reply:
I was debating that, I did it anyway, and kind of regretted it after. We should cache those values in language I'd think.